File: day-55-cf-863-d.py

n, num_requests, _ = map(int, raw_input().split(' '))

initial_values = map(int, raw_input().split(' '))

requests = []

for _ in xrange(num_requests):
    t, l, r = map(int, raw_input().split(' '))

    requests.append((t == 1 and 'right_shift' or 'reverse', l - 1, r - 1, ))

positions = [int(x) - 1 for x in raw_input().split(' ')]

# Instead of performing the sequence of requests on the array elements, play the sequence back in reverse order on
# the postitions of interest. The resulting positions will be the positions in the initial array from which the elements
# at positions of interest would have originated after the sequence would've been applied.
#
# This is possible because both subarray reverse and subarray shift are bijective operations.

for request_type, l, r in reversed(requests):
    if request_type == 'reverse':
        for i, p in enumerate(positions):
            # The inverse of reverse is just reverse
            if l <= p <= r:
                positions[i] = l + r - p
    elif request_type == 'right_shift':
        for i, p in enumerate(positions):
            # Perform left shift -- the inverse of right shift
            if p == l:
                positions[i] = r
            elif l < p <= r:
                positions[i] -= 1

print ' '.join(str(initial_values[p]) for p in positions)

File: day-55-cf-863-d.cpp
#include <iostream>
#include <vector>

struct request_t
{
    enum request_type { right_shift, reverse };

    request_type t;
    int l, r;
};

int main()
{
    std::ios_base::sync_with_stdio(false);

    int n, num_requests, num_positions;

    std::cin >> n >> num_requests >> num_positions;

    std::vector<int> initial_values((size_t)n);

    for(int i = 0; i < n; i++) {
        std::cin >> initial_values[i];
    }

    std::vector<request_t> requests((size_t)num_requests);

    for(int i = 0; i < num_requests; i++) {
        int t, l, r;

        std::cin >> t >> l >> r;

        requests[i] = {t == 1 ? request_t::request_type::right_shift : request_t::request_type::reverse, l, r};
    }

    std::vector<int> positions((size_t)num_positions);

    for(int i = 0; i < num_positions; i++) {
        std::cin >> positions[i];
    }


}
